src/rust/engine/grpc_util/src/lib.rs

```diff
@@ -124,47 +134,32 @@ pub fn create_tls_config(root_ca_certs: Option<Vec<u8>>) -> Result<ClientConfig,
   Ok(tls_config)
 }
 
-pub fn headers_to_metadata_map(headers: &BTreeMap<String, String>) -> Result<MetadataMap, String> {
-  let mut metadata_map = MetadataMap::with_capacity(headers.len());
-  for (key, value) in headers {
-    let key_ascii = AsciiMetadataKey::from_str(key.as_str()).map_err(|_| {
-      format!(
-        "Header key `{}` must be an ASCII value (as required by gRPC).",
-        key
-      )
-    })?;
-    let value_ascii = AsciiMetadataValue::from_str(value.as_str()).map_err(|_| {
-      format!(
-        "Header value `{}` for key `{}` must be an ASCII value (as required by gRPC).",
-        value, key
-      )
-    })?;
-    metadata_map.insert(key_ascii, value_ascii);
+pub fn headers_to_http_header_map(headers: &BTreeMap<String, String>) -> Result<HeaderMap, String> {
+  let http_headers = headers
+    .iter()
+    .map(|(key, value)| {
+      let header_name = HeaderName::from_str(key.as_str())
+        .map_err(|err| format!("Invalid header name {}: {}", key, err))?;
+
+      let header_value = HeaderValue::from_str(value.as_str())
+        .map_err(|err| format!("Invalid header value {}: {}", value, err))?;
+
+      Ok((header_name, header_value))
+    })
+    .collect::<Vec<Result<(HeaderName, HeaderValue), String>>>();
```

Comment:
If you're happy to only return the first error, rather than wanting to collect them all, you can collect directly to `Result<Vec<_>, _>` rather than `Vec<Result<_, _>>` which would allow you to avoid needing to then partition as well. And then actually, a `HeaderMap` is directly collectable-to; i.e. this function could literally be:

```rust
pub fn headers_to_http_header_map(headers: &BTreeMap<String, String>) -> Result<HeaderMap, String> {
  return headers
    .iter()
    .map(|(key, value)| {
      let header_name = HeaderName::from_str(key.as_str())
        .map_err(|err| format!("Invalid header name {}: {}", key, err))?;

      let header_value = HeaderValue::from_str(value.as_str())
        .map_err(|err| format!("Invalid header value {}: {}", value, err))?;

      Ok((header_name, header_value))
    })
    .collect::<Result<_, String>>()
    .map_err(|err| format!("header conversion error: {}", err));
}
```

Not saying you necessarily should, just letting you know that you _could_ :)